File: 2019/02/intcode.py
import sys

from typing import NamedTuple, List
from typing_extensions import Protocol

Program = List[int]

class HaltException(Exception):
    pass

class OpCode(Protocol):
    def eval(self, program: Program):
        ...

    @property
    def ip_step(self) -> int:
        ...

class Add(NamedTuple):
    src1: int
    src2: int
    dst: int

    def eval(self, program: Program):
        program[self.dst] = program[self.src1] + program[self.src2]

    @property
    def ip_step(self) -> int:
        return 4

class Mul(NamedTuple):
    src1: int
    src2: int
    dst: int

    def eval(self, program: Program):
        program[self.dst] = program[self.src1] * program[self.src2]

    @property
    def ip_step(self) -> int:
        return 4

class Halt(NamedTuple):
    def eval(self, program: Program):
        raise HaltException

    @property
    def ip_step(self) -> int:
        return 1

def parse(program: Program, ip: int) -> OpCode:
    op = program[ip]
    if op == 1:
        return Add(*program[ip + 1:ip + 4])
    if op == 2:
        return Mul(*program[ip + 1:ip + 4])
    if op == 99:
        return Halt()
    raise ValueError(f'Unknown opcode {op}')

def execute(program: Program) -> int:
    ip = 0
    while True:
        try:
            opcode = parse(program, ip)
            opcode.eval(program)
            ip += opcode.ip_step
        except HaltException:
            return program[0]

def problem1(program: Program):
    program[1] = 12
    program[2] = 2
    res = execute(program)
    print(f'final state: {program}')
    print(res)

def problem2(program: Program):
    for noun in range(100):
        for verb in range(100):
            this_prog = list(program)
            this_prog[1] = noun
            this_prog[2] = verb
            try:
                if execute(this_prog) == 19690720:
                    print(100 * noun + verb)
                    return
            except:
                pass
    print('No solution found', file=sys.stderr)

def main(args: List[str]) -> int:
    if len(args) != 2:
        print(f'Usage: {args[0]} 1|2 <input.txt', file=sys.stderr)
        return 0

    program = [int(x) for x in sys.stdin.readline().rstrip().split(',')]

    if args[1] == '1':
        problem1(program)
    elif args[1] == '2':
        problem2(program)
    else:
        print(f'Usage: {args[0]} 1|2 <input.txt', file=sys.stderr)

    return 0

if __name__ == '__main__':
    sys.exit(main(sys.argv))

File: 2019/02/intcode.rs
use std::{
    io,
    convert::TryInto,
};

#[derive(Debug, Copy, Clone)]
enum IntCodeError {
    ParseError,
    OutOfProgram,
    UnknownOpCode(i32),
    InvalidAddress(i32),
}

#[derive(Debug, Copy, Clone)]
enum OpCode {
    Add(usize, usize, usize),
    Mul(usize, usize, usize),
    Halt,
}

#[inline]
fn indirect_address(program: &mut [i32], ptr: usize)
  -> Result<usize, IntCodeError> {
    let addr = *program.get(ptr).ok_or(IntCodeError::OutOfProgram)?;
    addr.try_into().map_err(|_| IntCodeError::InvalidAddress(addr))
}

impl OpCode {
    fn parse(program: &mut [i32], ip: usize)
      -> Result<(OpCode, usize), IntCodeError> {
        match *program.get(ip).ok_or(IntCodeError::OutOfProgram)? {
            1 => {
                let opcode = OpCode::Add(
                    indirect_address(program, ip + 1)?,
                    indirect_address(program, ip + 2)?,
                    indirect_address(program, ip + 3)?,
                );
                Ok((opcode, ip + 4))
            },

            2 => {
                let opcode = OpCode::Mul(
                    indirect_address(program, ip + 1)?,
                    indirect_address(program, ip + 2)?,
                    indirect_address(program, ip + 3)?,
                );
                Ok((opcode, ip + 4))
            },

            99 => Ok((OpCode::Halt, ip)),

            n => Err(IntCodeError::UnknownOpCode(n)),
        }
    }

    fn eval(&self, program: &mut [i32]) -> Result<(), IntCodeError> {
        match *self {
            OpCode::Add(src1, src2, dst) => {
                let arg1 = *program.get(src1)
                    .ok_or(IntCodeError::InvalidAddress(src1 as i32))?;
                let arg2 = *program.get(src2)
                    .ok_or(IntCodeError::InvalidAddress(src2 as i32))?;
                let dst = program.get_mut(dst)
                    .ok_or(IntCodeError::InvalidAddress(dst as i32))?;
                *dst = arg1 + arg2;
            },

            OpCode::Mul(src1, src2, dst) => {
                let arg1 = *program.get(src1)
                    .ok_or(IntCodeError::InvalidAddress(src1 as i32))?;
                let arg2 = *program.get(src2)
                    .ok_or(IntCodeError::InvalidAddress(src2 as i32))?;
                let dst = program.get_mut(dst)
                    .ok_or(IntCodeError::InvalidAddress(dst as i32))?;
                *dst = arg1 * arg2;
            },

            OpCode::Halt => {},
        }

        Ok(())
    }
}

fn execute(program: &mut [i32]) -> Result<(), IntCodeError> {
    let mut ip = 0;
    loop {
        let (opcode, next_ip) = OpCode::parse(program, ip)?;
        ip = next_ip;
        match opcode {
            OpCode::Halt => return Ok(()),
            _ => opcode.eval(program)?,
        }
    }
}

fn problem1(program: &mut [i32]) -> Result<(), IntCodeError> {
    program[1] = 12;
    program[2] = 2;
    execute(program)?;
    println!("final state: {:?}", program);
    Ok(())
}

fn main() -> Result<(), IntCodeError> {
    let mut program = String::new();
    io::stdin().read_line(&mut program)
        .expect("failed to read program from stdin");
    let mut program: Vec<i32> = program.trim_end().split(',')
        .map(|word| word.parse().map_err(|_| IntCodeError::ParseError))
        .collect::<Result<_, _>>()?;

    problem1(&mut program[..])
}
